onnx/backend/test_rnn.py

```diff
@@ -0,0 +1,318 @@
+# Provides reference implementation of CuDNN-style RNNs in numpy
+
+import numpy as np
+from itertools import chain, izip
+from collections import defaultdict
+from .test_util import N
+
+
+def interleave(x, y):
+    return list(chain.from_iterable(izip(x, y)))
+
+
+class RNNBase(object):
+    def param_size(self):
+        return sum(p.size for p in self.all_params())
+
+    def all_params(self):
+        return self.all_weights() + self.all_biases()
+
+    def concat_params(self):
+        return np.concatenate(tuple(x.flatten() for x in self.all_params()))
+
+    def feed_params(self, w):
+        assert self.param_size() == w.size, \
+            '%d vs %d' % (self.param_size(), w.size)
+        offset = 0
+        for x in self.all_params():
+            x.ravel()[:] = w[offset:offset + x.size]
+            offset += x.size
+        return self
+
+
+# TODO: add support for it in C2
+def ReluCell(h, c, hh, ih):
+    s = hh + ih
+    return np.maximum(s[:, 0, :], 0), None
+
+
+ReluCell.GATES = 1
+ReluCell.HAS_CELL = False
+
+
+def TanhCell(h, c, hh, ih):
+    s = hh + ih
+    return np.tanh(s[:, 0, :]), None
+
+
+TanhCell.GATES = 1
+TanhCell.HAS_CELL = False
+
+
+def GRUCell(h, c, hh, ih):
+    RESET = 0
+    INPUT = 1
+    HIDDEN = 2
+    s = hh + ih
+    s[:, [RESET, INPUT], :] = 1.0 / (1.0 + np.exp(-s[:, [RESET, INPUT], :]))
+    q = np.tanh(s[:, RESET, :] * hh[:, HIDDEN, :] + ih[:, HIDDEN, :])
+    return (1 - s[:, INPUT, :]) * q + s[:, INPUT, :] * h, None
+
+
+GRUCell.GATES = 3
+GRUCell.HAS_CELL = False
+
+
+def LSTMCell(h, c, hh, ih):
+    INPUT = 0
+    FORGET = 1
+    CELL = 2
+    OUTPUT = 3
+    s = hh + ih
+    s[:, [INPUT, FORGET, OUTPUT], :] = 1.0 / (
+        1.0 + np.exp(-s[:, [INPUT, FORGET, OUTPUT], :]))
+    s[:, CELL, :] = np.tanh(s[:, CELL, :])
+    c = s[:, FORGET, :] * c + s[:, INPUT, :] * s[:, CELL, :]
+    h = s[:, OUTPUT, :] * np.tanh(c)
+    return h, c
+
+
+LSTMCell.GATES = 4
+LSTMCell.HAS_CELL = True
+
+
+class RNNRef(RNNBase):
+    IH = 'ih'
+    HH = 'hh'
+
+    def __init__(self, cell, hidden_size, dim_in, skip=False):
+        if skip:
+            assert hidden_size == dim_in
+        self.cell = cell
+        self.hidden_size = hidden_size
+        self.dim_in = dim_in
+        self.weights = {}
+        self.biases = {}
+        if not skip:
+            self.weights[self.IH] = np.random.randn(
+                cell.GATES, self.hidden_size, self.dim_in).astype(np.float32)
+        else:
+            self.weights[self.IH] = np.empty((0))
```

Comment:
nit: missed a `,` in the tuple, but `numpy.empty` can take single int as shape too.